File: binary_alphabet/rnd_dbseq.cpp
#include<stdio.h>
#include<stdlib.h>
#include<time.h>
#include<vector>
#include"unionfind.hpp"

using namespace std;

#define ll unsigned long long int

int k;
ll m, n;
int num_cycles;
int *cycle;
bool *bwt_enc;

//--------------------------------------------------------------------------------
// Random bitstring representing the encoded bwt
//--------------------------------------------------------------------------------
void rnd_bistr(){
    bwt_enc[0] = bwt_enc[m-1] = 1;
    for (size_t i = 1; i < m-1; i++)  bwt_enc[i] = rand() % 2;
}

//--------------------------------------------------------------------------------
// Assign cycle to the permutation of the bwt_enc of a dB set
//--------------------------------------------------------------------------------
void bwt2cycle() {
    ll i = 0;
    num_cycles = 1;
    while (i < n) {
        while (!cycle[i]) {
            cycle[i] = num_cycles;
            i = i/2 + (bwt_enc[i/2] ^ (i%2))*m;
        }
        i++;
        if (i<n && !cycle[i]) num_cycles++;
    }
}

//--------------------------------------------------------------------------------
// Return the inverse of the bwt_enc of de Bruijn sequence
//--------------------------------------------------------------------------------
void inv_bwt() {
    ll i = 0;
    do {
        printf("%d", bwt_enc[i/2]^int(i%2)); // dbseq with numbers {0,1}
        //printf("%c", 'a'+(bwt_enc[i/2]^int(i%2))); // dbseq with letters {a,b}
        i = i/2 + (bwt_enc[i/2] ^ (i%2)) * m;
    } while(i != 0);
    printf("\n");
}

//--------------------------------------------------------------------------------
// Generate a random binary de Bruijn sequence of order k
//--------------------------------------------------------------------------------
void rnd_debruijn() {
    rnd_bistr();
    bwt2cycle();

    if (num_cycles == 1) {
        inv_bwt();
        return;
    }

    // Possible edges of the cycle graph
    vector<ll> edges;
    for (ll i = 0; i < n; i+=2) 
        if (cycle[i] != cycle[i+1]) edges.push_back(i);

    // Generate a spanning tree of the cycle graph
    UnionFind ST = UnionFind(num_cycles);
    int num_edges = edges.size();

    while (num_cycles > 1) {
        int j = rand() % num_edges--;

        ll e = edges[j];
        edges[j] = edges[num_edges];
        edges[num_edges] = e;
        // If the edge can be added to the ST do it by toggling the bit in position e/2
        if (ST.Union(cycle[e], cycle[e+1])) {
            bwt_enc[e/2]=!bwt_enc[e/2];
            num_cycles--;
        }
    }

    delete [] cycle;
    
    inv_bwt();
}

//--------------------------------------------------------------------------------
// MAIN
//--------------------------------------------------------------------------------
int main(int argc, char **argv) {
    srand(time(NULL));

    if (argc > 1){
        k = atoi(argv[1]);
    } else {
        k = 4;
    }

    m = 1<<(k-1);
    n = m*2;
    bwt_enc = new bool[m];
    cycle = new int[n]();

    rnd_debruijn();

    delete [] bwt_enc;
 
    return 0;
}

File: binary_alphabet/python/rnd_dbseq.py
import random 
import sys
from itertools import product
from unionfind import UnionFind

Sig = ['0','1']
#Sig = ['a','b']

#--------------------------------------------------------------------------------
# Random bitstring representing the encoded bwt
#--------------------------------------------------------------------------------
def rnd_bitstr(m):
    bwt_enc = [0]*m
    for i in range(m):
        bwt_enc[i] = random.randint(0,1)
    return bwt_enc

#--------------------------------------------------------------------------------
# Assign cycle to the permutation of the bwt_enc of a dB set
#--------------------------------------------------------------------------------
def bwt2cycle(bwt_enc, n, m):
    cycle = [0]*n

    i = 0
    num_cycles = 1
    while i < n:
        while cycle[i] == 0:
            cycle[i] = num_cycles
            i = i//2 + (bwt_enc[i//2] ^ (i%2))*m
        i+=1
        if i<n and cycle[i]==0: num_cycles+=1

    return [cycle, num_cycles]

#--------------------------------------------------------------------------------
# Return the inverse of the bwt_enc of de Bruijn sequence
#--------------------------------------------------------------------------------
def inv_bwt(bwt_enc, m):
    i = 0
    dbseq = Sig[bwt_enc[i]]
    i = bwt_enc[0]*m
    while i != 0:
        dbseq += Sig[bwt_enc[i//2]^i%2]  # print db seq with {a,b}
        i = i//2 + (bwt_enc[i//2]^(i%2))*m
    return dbseq

#--------------------------------------------------------------------------------
# Generate a random binary de Bruijn sequence of order k
#--------------------------------------------------------------------------------
def rnd_debruijn(k):
    m = 2**(k-1)
    n = m*2
    bwt_enc = rnd_bitstr(m)
    [cycle, num_cycles] = bwt2cycle(bwt_enc, n, m)
    
    if num_cycles == 1: 
        return inv_bwt(bwt_enc, m)

    edges = []
    for i in range(0,n,2):
        if cycle[i] != cycle[i+1]: edges.append(i)

    ST = UnionFind(num_cycles)
    num_edges = len(edges)

    while num_cycles > 1:
        j = random.randint(0, num_edges-1)
        num_edges -= 1

        e = edges[j]
        edges[j]= edges[num_edges]
        edges[num_edges] = e

        if ST.union(cycle[e], cycle[e+1]):
            bwt_enc[e//2] = not bwt_enc[e//2]
            num_cycles-=1

    return inv_bwt(bwt_enc, m)

#--------------------------------------------------------------------------------
# Check for de Bruijn sequence
#--------------------------------------------------------------------------------
def is_debruijn(s, Sig, k):
    d = dict()
    if len(s) == len(Sig)**k:
        # initialize dict of all kmers with 0s
        for i in product(Sig, repeat=k):
            d[''.join(i)] = 0
        # Control the first kmers until
        for i in range(len(s)):
            j = (i+k)%len(s)
            if j < i:
                kmer = s[i:] + s[:j]
            else:
                kmer = s[i:j]

            if d[kmer] == 0:
                d[kmer] += 1
            else:
                return False
    else:
        return False

    return True

#--------------------------------------------------------------------------------
# MAIN
#--------------------------------------------------------------------------------
def main():
    #random.seed(1479)
    if len(sys.argv) >= 2:
        k = int(sys.argv[1])
    else:
        k = 6
    dbseq = rnd_debruijn(k)
    print(dbseq)
    #print(is_debruijn(dbseq, Sig, k))

if __name__ == "__main__":
    main()
